core/state/pruner/pruner.go

```diff
@@ -266,6 +269,15 @@ func (p *Pruner) Prune(root common.Hash) error {
 	// state is picked for usage.
 	deleteCleanTrieCache(p.trieCachePath)
 
+	// All the state roots of the middle layer should be forcibly pruned,
+	// otherwise the dangling state will be left.
+	middleRoots := make(map[common.Hash]struct{})
+	for _, layer := range layers {
+		if layer.Root() == root {
+			break
+		}
+		middleRoots[layer.Root()] = struct{}{}
+	}
```

Comment:
Another way to do it, which doesn't rely on all snapshot layers to exist, would be to just iterate backwards on the blocks

Reply:
Yes, I know. But now we have the layers and we can ensure that the target layer exists in these layers. So we don't need the db access.